packages/react/src/components/Notification/Notification.js

```diff
@@ -266,25 +267,33 @@ export function ToastNotification({
     [`${prefix}--toast-notification--${kind}`]: kind,
   });
 
+  const handleClose = useCallback(
+    (evt) => {
+      if (!onClose || onClose(evt) !== false) {
+        setIsOpen(false);
+      }
+    },
+    [onClose]
+  );
+
   function handleCloseButtonClick(event) {
-    setIsOpen(false);
     onCloseButtonClick(event);
+    handleClose(event);
   }
 
   useEffect(() => {
     if (!timeout) {
       return;
     }
 
-    const timeoutId = window.setTimeout(() => {
-      setIsOpen(false);
-      onCloseButtonClick(event);
+    const timeoutId = window.setTimeout((evt) => {
+      handleClose(evt);
     }, timeout);
 
     return () => {
       window.clearTimeout(timeoutId);
     };
-  }, [onCloseButtonClick, timeout]);
+  }, [handleClose, timeout]);
```

Comment:
This might be related to the `useCallback` note above, but if we save the `onClose` prop in a ref we might be able to do:

```js
function ToastNotification({
  // ...
  onClose
}) {
  const savedOnClose = useRef(onClose);

  useEffect(() => {
    savedOnClose.current = onClose;
  });

  useEffect(() => {
    if (!timeout) {
      return;
    }

    const timeoutId = window.setTimeout((event) => {
      setIsOpen(false);
      if (savedOnClose.current) {
        savedOnClose.current(event);
      }
    });
    return () => {
      window.clearTimeout(timeoutId);
    };
  }, [timeout]);
}
```

Reply:
makes sense together with the previous comment, updated